cs-multi-shell/src/app/authentication.service.ts

```diff
@@ -0,0 +1,131 @@
+import { Injectable } from '@angular/core';
+import { Capacitor } from '@capacitor/core';
+import { AuthConnect, AuthResult, OktaProvider, ProviderOptions, TokenType } from '@ionic-enterprise/auth';
+import { Platform } from '@ionic/angular';
+import { urlSchemeIonicAuthOptions, webIonicAuthOptions } from '../environments/environment';
+import { RouteService } from './route.service';
+import { VaultService } from './vault.service';
+
+@Injectable({ providedIn: 'root' })
+export class AuthenticationService {
+    private result: AuthResult | undefined;
+
+    constructor(
+        private routeService: RouteService,
+        private vaultService: VaultService) {
+        this.init();
+    }
+
+    /**
+     * Initialize: setup Auth Connect, and get token from storage if available
+     */
+    public async init() {
+        await AuthConnect.setup({
+            platform: Capacitor.getPlatform() == 'web' ? 'web' : 'capacitor',
+            logLevel: 'NONE',
+            ios: {
+                webView: 'private',
+                safariWebViewOptions: {
+                    dismissButtonStyle: 'close',
+                    preferredBarTintColor: '#FFFFFF',
+                    preferredControlTintColor: '#333333'
+                }
+            },
+            android: {
+                isAnimated: false,
+                showDefaultShareMenuItem: false
+            },
+            web: {
+                uiMode: 'current',
+                authFlow: 'PKCE'
+            }
+        });
+
+        try {
+            this.result = await this.vaultService.get();
+        } catch (error) {
+            console.error(error);
+            this.result = undefined;
+        }
+    }
+
+    /**
+     * Login
+     */
+    public async login() {
+        this.result = await AuthConnect.login(this.oktaProvider(), this.getAuthOptions());
+        await this.vaultService.set(this.result);
+        this.routeService.goToRoot();
+    }
+
+    /**
+     * Called for the web platform. Passes Auth Connect the auto info from query parameters
+     * to get the auth object which we store and redirect to the home page
+     */
+    public async handleLogin() {
+        const urlParams = new URLSearchParams(window.location.search);
```

Comment:
This is needed just for web, so that the login can occur in the same window, rather than a popup window (which is often blocked by browsers).